### DISCUSSION TITLE
how to block page rendering until data is ready

### DISCUSSION THREAD

**Discussion by ghost** [nuxt/nuxt#17006] [Questions] [answered]
I'm trying to load data on the server before the page renders, for that I'm using `useAsyncData` + `pinia`

```vue
<script setup lang="ts">
import { useBudgetStore } from '@/stores/budget'

const budgetStore = useBudgetStore()

useAsyncData('budgets', budgetStore.fetchBudgets)
useAsyncData('overview', budgetStore.fetchOverview)
</script>
```

It works perfectly when we directly access this page, ex `/budget`, the page will only be displayed after the data has been loaded and filled in the store

but if there is a navigation between routes, ex `/login` to `/budget` using `navigateTo()`, `router.push()` or `<NuxtLink>`, 

the `useAsyncData` will try to perform the request on the client side and it won't block the page loading which will result in a lot of errors as the components depend on the data

I already tried putting `await`
```ts
await useAsyncData('budgets', budgetStore.fetchBudgets)
await useAsyncData('overview', budgetStore.fetchOverview)
```

but the rendering page is released after the completion of the first `useAsyncData`, and the second `useAsyncData runs from both, the client and the server side

**Comment by iPrytz** [comment, marked as answer]
Here is our code that does work as you explain. Make sure `fetchBudgets `and `fetchOverview` are returning promises
```
import { useXStore, useYStore } from '@/stores'
const route = useRoute();
const reactiveQuery = computed(() => route.query.q);
const xStore = useXStore()
const yStore = useYStore()

const refresh = async () => {
  return await Promise.all([xStore.fetchX(), yStore.fetchY()])
}

watch(reactiveQuery, await refresh);
await useAsyncData('x-y', () => refresh())
```

**Reply by ghost** [reply]
works perfectly, thank you very much!
